### PR TITLE
WIP on initializing NE database [#108]

### PR DESCRIPTION
**Pull request by bdon**
@nvkelso this is a proof of concept, discovered some issues - will leave inline comments

### CODE REVIEW

File: tiles/src/main/java/com/protomaps/basemap/Basemap.java [hunk 1]
```diff
@@ -116,20 +118,28 @@ static void run(Arguments args) throws Exception {
...
-    planetiler.addShapefileSource("osm_water", sourcesDir.resolve("water-polygons-split-3857.zip"),
-      "https://osmdata.openstreetmap.de/download/water-polygons-split-3857.zip")
+      .addShapefileSource("osm_water", sourcesDir.resolve("water-polygons-split-3857.zip"),
+        "https://osmdata.openstreetmap.de/download/water-polygons-split-3857.zip")
       .addShapefileSource("osm_land", sourcesDir.resolve("land-polygons-split-3857.zip"),
-        "https://osmdata.openstreetmap.de/download/land-polygons-split-3857.zip")
-      .run();
+        "https://osmdata.openstreetmap.de/download/land-polygons-split-3857.zip");
+
+    Downloader.create(planetiler.config(), planetiler.stats()).add("ne", neUrl, nePath).run();
```
Comment (bdon) [inline tiles/src/main/java/com/protomaps/basemap/Basemap.java:136]: A little awkward, but this lets us share the same DB file that we use for lookups and we use for actual features. We load the DB twice, which isn't expensive.

File: tiles/src/main/java/com/protomaps/basemap/Basemap.java [hunk 2]
```diff
@@ -116,20 +118,28 @@ static void run(Arguments args) throws Exception {
...
+
+    var tmpDir = nePath.resolveSibling(nePath.getFileName() + "-unzipped");
+    var naturalEarthDb = NaturalEarthDb.fromSqlite(nePath, tmpDir);
+    System.out.println(naturalEarthDb.getByWikidataId("Q60"));
```
Comment (bdon) [inline tiles/src/main/java/com/protomaps/basemap/Basemap.java:140]: This is how we look up by a Wikidata ID

File: tiles/src/main/java/com/protomaps/basemap/Basemap.java [hunk 3]
```diff
@@ -116,20 +118,28 @@ static void run(Arguments args) throws Exception {
...
+        "https://osmdata.openstreetmap.de/download/water-polygons-split-3857.zip")
       .addShapefileSource("osm_land", sourcesDir.resolve("land-polygons-split-3857.zip"),
-        "https://osmdata.openstreetmap.de/download/land-polygons-split-3857.zip")
-      .run();
+        "https://osmdata.openstreetmap.de/download/land-polygons-split-3857.zip");
+
+    Downloader.create(planetiler.config(), planetiler.stats()).add("ne", neUrl, nePath).run();
+
+    var tmpDir = nePath.resolveSibling(nePath.getFileName() + "-unzipped");
+    var naturalEarthDb = NaturalEarthDb.fromSqlite(nePath, tmpDir);
```
Comment (bdon) [inline tiles/src/main/java/com/protomaps/basemap/Basemap.java:139]: We'll eventually consolidate on Geopackage...

File: tiles/src/main/java/com/protomaps/basemap/feature/NaturalEarthDb.java [hunk 4]
```diff
@@ -0,0 +1,90 @@
...
+
+  public NePopulatedPlace getByWikidataId(String wikidataId) {
+    return this.placesByWikidataId.get(wikidataId);
+  }
+
+  public static NaturalEarthDb fromSqlite(Path path, Path unzippedDir) {
+    boolean keepUnzipped = true;
+    List<NePopulatedPlace> results = new ArrayList<>();
+    try {
+      Path extracted;
```
Comment (bdon) [inline tiles/src/main/java/com/protomaps/basemap/feature/NaturalEarthDb.java:48]: Most of this is ugly cruft we can refactor out later.

File: tiles/src/main/java/com/protomaps/basemap/feature/NaturalEarthDb.java [hunk 5]
```diff
@@ -0,0 +1,90 @@
...
+    }
+  }
+
+  private Map<String, NePopulatedPlace> placesByWikidataId;
+
+  public record NePopulatedPlace(String name, String wikidataId, String minZoom, String populationRank) {};
+
+  private NaturalEarthDb(List<NePopulatedPlace> places) {
+    this.placesByWikidataId = places.stream().filter(place -> place.wikidataId != null)
+      .collect(Collectors.toMap(place -> place.wikidataId, place -> place, (s, a) -> s)); // TODO conflicts
```
Comment (bdon) [inline tiles/src/main/java/com/protomaps/basemap/feature/NaturalEarthDb.java:37]: @nvkelso there are instances in ne_10m_populated_places of multiple rows with the same Wikidata ID. Which one should win?

Comment (nvkelso) [reply tiles/src/main/java/com/protomaps/basemap/feature/NaturalEarthDb.java:37]: The one with larger population (`pop_max`) value?

Let me dig into that a bit...

File: tiles/src/main/java/com/protomaps/basemap/feature/NaturalEarthDb.java [hunk 6]
```diff
@@ -0,0 +1,183 @@
...
+    this.admin0sByWikidataId = countries.stream().filter(c -> c.wikidataId != null)
+      .collect(Collectors.toMap(c -> c.wikidataId, c -> c, (c1, c2) -> c1.minLabel() < c2.minLabel() ? c1 : c2));
+
+    this.admin1sByIso31662 = statesProvinces.stream()
+      .collect(Collectors.toMap(s -> s.iso3166_2, s -> s, (s1, s2) -> s1.minLabel() < s2.minLabel() ? s1 : s2));
+
+    this.admin1sByWikidataId = statesProvinces.stream().filter(s -> s.wikidataId != null)
+      .collect(Collectors.toMap(s -> s.wikidataId, s -> s));
+
+    // resolve wikidata conflicts by choosing the larger rankMax
```
Comment (nvkelso) [inline tiles/src/main/java/com/protomaps/basemap/feature/NaturalEarthDb.java:73]: Nice!

File: tiles/src/main/java/com/protomaps/basemap/layers/Pois.java [hunk 7]
```diff
@@ -51,6 +58,11 @@ public void processFeature(SourceFeature sf, FeatureCollector features) {
       String kindDetail = "";
       Integer minZoom = 15;
 
+      String wikidata = sf.getString("wikidata");
+      if (wikidata != null) {
+        qrankDb.get(wikidata);
```
Comment (nvkelso) [inline tiles/src/main/java/com/protomaps/basemap/layers/Pois.java:63]: TODO: More work needed to influence a feature's minZoom?

**Pull request merged**